### PR TITLE
Teach build-script to build Foundation Dependencies (Zlib, libxml2, LibCurl)

### PR DESCRIPTION
**Pull request by etcwilde**
This PR teaches build-script to build `libxml2`, `curl`, and `zlib`, which are dependencies for building foundation.

### CODE REVIEW

**Comment by etcwilde** [comment]
@swift-ci please test

File: utils/build-script-impl [hunk 1]
```diff
@@ -2516,6 +2520,40 @@ for host in "${ALL_HOSTS[@]}"; do
                     LIBICU_BUILD_ARGS=()
                 fi
 
+                if [[ ! "${SKIP_BUILD_LIBXML2}" ]]; then
+                    BASE_INSTALL_DIR="$(get_host_install_destdir ${host})"
+                    LIBXML2_HEADERS="${BASE_INSTALL_DIR}/usr/include/libxml2"
+                    LIBXML2_BUILD_ARGS=(-DLIBXML2_INCLUDE_DIR:PATH="$LIBXML2_HEADERS")
+                    # Find static library
+                    LIBXML2_LIB=$(find "${BASE_INSTALL_DIR}"  -name 'libxml2.a')
```
Comment (compnerd) [inline utils/build-script-impl:2528]: Please don't use `find` for this.  You know it should be at `${BASE_INSTALL_DIR}/usr/lib/libxml2.a`, just write it out.

File: utils/build-script-impl [hunk 2]
```diff
@@ -2516,6 +2520,40 @@ for host in "${ALL_HOSTS[@]}"; do
...
+                    if [[ ! "${LIBXML2_LIB}" ]]; then
```
Comment (compnerd) [inline utils/build-script-impl:2529]: Rather than the stringy test, `-e` should be good.

File: utils/build-script-impl [hunk 3]
```diff
@@ -2516,6 +2520,40 @@ for host in "${ALL_HOSTS[@]}"; do
...
+                        LIBXML2_LIB=$(find "${BASE_INSTALL_DIR}" -name 'libxml2.so' -o -name 'libxml2.dll')
```
Comment (compnerd) [inline utils/build-script-impl:2530]: `find` won't work for Windows, so the `.dll` is meaningless.  Additionally, the static library name is incorrect for windows.

File: utils/build-script-impl [hunk 4]
```diff
@@ -2516,6 +2520,40 @@ for host in "${ALL_HOSTS[@]}"; do
...
+                    fi
+                    LIBXML2_BUILD_ARGS+=(-DLIBXML2_LIBRARY:PATH="$LIBXML2_LIB")
+                else
+                    LIBXML2_BUILD_ARGS=()
+                fi
+
+                if [[ ! "${SKIP_BUILD_ZLIB}" ]]; then
+                    BASE_INSTALL_DIR="$(get_host_install_destdir ${host})"
+                    ZLIB_LIB=$(find $BASE_INSTALL_DIR -name 'libz.a' -o \
+                               -name 'libz.so' -o -name 'libz.dll' | head -n 1)
```
Comment (compnerd) [inline utils/build-script-impl:2543]: `find` wont work for windows, `libz.a` is the wrong name (`zlibstatic.lib` is the library on Windows).  I think that we should do the same as windows here - always statically link zlib.

File: utils/build-script-impl [hunk 5]
```diff
@@ -2516,6 +2520,40 @@ for host in "${ALL_HOSTS[@]}"; do
...
+                    LIBXML2_BUILD_ARGS=(-DLIBXML2_INCLUDE_DIR:PATH="$LIBXML2_HEADERS")
+                    # Find static library
+                    LIBXML2_LIB=$(find "${BASE_INSTALL_DIR}"  -name 'libxml2.a')
+                    if [[ ! "${LIBXML2_LIB}" ]]; then
+                        LIBXML2_LIB=$(find "${BASE_INSTALL_DIR}" -name 'libxml2.so' -o -name 'libxml2.dll')
+                        LIBXML2_BUILD_ARG+=($LIBXML2_LIB)
+                    else
+                        LIBXML2_BUILD_ARG+=(-DLIBXML2_DEFINITIONS="-DLIBXML_STATIC")
+                    fi
+                    LIBXML2_BUILD_ARGS+=(-DLIBXML2_LIBRARY:PATH="$LIBXML2_LIB")
```
Comment (compnerd) [inline utils/build-script-impl:2535]: I think that we should do the same thing as windows here - always statically link

File: utils/build_swift/build_swift/driver_arguments.py [hunk 6]
```diff
@@ -709,6 +718,31 @@ def create_argument_parser():
     option('--libicu', toggle_true('build_libicu'),
            help='build libicu')
 
+    option('--libxml2', toggle_true('build_libxml2'),
+           default=False,
+           help='build libxml2')
```
Comment (compnerd) [inline utils/build_swift/build_swift/driver_arguments.py:723]: I think we should just drop this

File: utils/build_swift/build_swift/driver_arguments.py [hunk 7]
```diff
@@ -709,6 +718,31 @@ def create_argument_parser():
...
+
+    option('--static-libxml2', toggle_true(['build_libxml2',
+                                            'build_static_libxml2']),
+           default=False,
+           help='build static libxml2')
+
+    option('--zlib', toggle_true('build_zlib'),
+            default=False,
+            help='build zlib')
```
Comment (compnerd) [inline utils/build_swift/build_swift/driver_arguments.py:732]: Similar

File: utils/build_swift/build_swift/driver_arguments.py [hunk 8]
```diff
@@ -709,6 +718,31 @@ def create_argument_parser():
...
+
+    option('--static-zlib', toggle_true(['build_zlib', 'build_static_zlib']),
+           default=False,
+           help='build static zlib')
+
+    option('--curl', toggle_true('build_curl'),
+            default=False,
+            help="build curl libraries")
```
Comment (compnerd) [inline utils/build_swift/build_swift/driver_arguments.py:740]: Similar

File: utils/swift_build_support/swift_build_support/products/curl.py [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+        self.cmake_options.define('CURL_DISABLE_TFTP', 'YES')
+        self.cmake_options.define('CURL_DISABLE_RTSP', 'YES')
+        self.cmake_options.define('CURL_DISABLE_PROXY', 'YES')
+        self.cmake_options.define('CURL_DISABLE_POP3', 'YES')
+        self.cmake_options.define('CURL_DISABLE_IMAP', 'YES')
+        self.cmake_options.define('CURL_DISABLE_SMTP', 'YES')
+        self.cmake_options.define('CURL_DISABLE_GOPHER', 'YES')
+        self.cmake_options.define('CURL_ZLIB', 'YES')
+        self.cmake_options.define('ENABLE_UNIX_SOCKETS', 'NO')
+        self.cmake_options.define('ENABLE_THREADED_RESOLVER', 'NO')
```
Comment (compnerd) [inline utils/swift_build_support/swift_build_support/products/curl.py:115]: Does this match the windows configuration?

**Comment by buttaface** [comment]
Heh, interesting, this could certainly help with cross-compiling Foundation. Any reason you left [the libicu dependency](https://github.com/apple/swift-corelibs-foundation/blob/0920b409cf885f568602a198790edd78110c8c40/CMakeLists.txt#L50) out?

**Comment by etcwilde** [comment]
> Heh, interesting, this could certainly help with cross-compiling Foundation. Any reason you left [the libicu dependency](https://github.com/apple/swift-corelibs-foundation/blob/0920b409cf885f568602a198790edd78110c8c40/CMakeLists.txt#L50) out?

Indeed. `build-script-impl` already knows how to build libicu. Someone could come along and migrate it to a python product, but that's not my goal today anyway.
https://github.com/apple/swift/blob/ef2c999c7a8c9004ca107ce4242a055874e089e6/utils/build-script-impl#L2617-L2670

And it's hooked into foundation here;
https://github.com/apple/swift/blob/ef2c999c7a8c9004ca107ce4242a055874e089e6/utils/build-script-impl#L2495-L2516

**Comment by buttaface** [comment]
Oh, my mistake, forgot about that. 🤓

File: utils/swift_build_support/swift_build_support/products/curl.py (hunk 9 above)
Comment (etcwilde) [reply utils/swift_build_support/swift_build_support/products/curl.py:115]: If I flip `SCHANNEL` to `YES`, it should match https://github.com/apple/swift/blob/64e9d0b27f3b8f33a8b7c8533b215562bc97349b/utils/build-windows-toolchain.bat#L133-L153

Comment (compnerd) [reply utils/swift_build_support/swift_build_support/products/curl.py:115]: Yes, cannot enable that on non-Windows, that makes sense.

**Comment by etcwilde** [comment]
@swift-ci please test

**Comment by etcwilde** [comment]
@swift-ci please test

File: utils/build-script-impl (hunk 1 above)
Comment (etcwilde) [reply utils/build-script-impl]: Done

File: utils/build-script-impl (hunk 2 above)
Comment (etcwilde) [reply utils/build-script-impl]: Done

File: utils/build-script-impl (hunk 4 above)
Comment (etcwilde) [reply utils/build-script-impl]: Done

File: utils/build-script-impl (hunk 5 above)
Comment (etcwilde) [reply utils/build-script-impl]: Done

File: utils/build_swift/build_swift/driver_arguments.py (hunk 6 above)
Comment (etcwilde) [reply utils/build_swift/build_swift/driver_arguments.py]: Done

File: utils/build_swift/build_swift/driver_arguments.py (hunk 7 above)
Comment (etcwilde) [reply utils/build_swift/build_swift/driver_arguments.py]: Done

**Review by compnerd — APPROVED** [review]

**Comment by etcwilde** [comment]
@swift-ci please test

**Comment by etcwilde** [comment]
@swift-ci please test Linux

**Pull request merged**